src/lightning/data/streaming/resolver.py

```diff
@@ -0,0 +1,340 @@
+import re
+import sys
+from dataclasses import dataclass
+from pathlib import Path
+from time import sleep
+from typing import Optional, Union
+from urllib import parse
+
+from lightning_cloud.openapi import V1CloudSpace
+from lightning_cloud.rest_client import LightningClient
+
+# To avoid adding lightning_utilities as a dependency for now.
+try:
+    import boto3
+    import botocore
+
+    _BOTO3_AVAILABLE = True
+except Exception:
+    _BOTO3_AVAILABLE = False
+
+
+try:
+    from lightning_sdk import Machine, Studio
+
+    _LIGHTNING_SDK_AVAILABLE = True
+except (ImportError, ModuleNotFoundError):
+
+    class Machine:  # type: ignore
+        pass
+
+    _LIGHTNING_SDK_AVAILABLE = False
+
+
+@dataclass
+class Dir:
+    """Holds a directory path and possibly its associated remote URL."""
+
+    path: Optional[str] = None
+    url: Optional[str] = None
+
+
+def _resolve_dir(dir_path: Optional[Union[str, Dir]]) -> Dir:
+    if isinstance(dir_path, Dir):
+        return Dir(path=str(dir_path.path) if dir_path.path else None, url=str(dir_path.url) if dir_path.url else None)
+
+    if dir_path is None:
+        return Dir()
+
+    if not isinstance(dir_path, str):
+        raise ValueError(f"`dir_path` must be a `Dir` or a string, got: {dir_path}")
+
+    assert isinstance(dir_path, str)
+
+    if dir_path.startswith("s3://"):
+        return Dir(path=None, url=dir_path)
+
+    dir_path = _resolve_time_template(dir_path)
+
+    dir_path_absolute = str(Path(dir_path).absolute().resolve())
+
+    if dir_path_absolute.startswith("/teamspace/studios/this_studio"):
+        return Dir(path=dir_path_absolute, url=None)
+
+    if dir_path_absolute.startswith("/.project"):
+        dir_path_absolute = dir_path
+
+    if dir_path_absolute.startswith("/teamspace/studios") and len(dir_path_absolute.split("/")) > 3:
+        return _resolve_studio(dir_path_absolute, dir_path_absolute.split("/")[3], None)
+
+    if dir_path_absolute.startswith("/teamspace/s3_connections") and len(dir_path_absolute.split("/")) > 3:
+        return _resolve_s3_connections(dir_path_absolute)
+
+    if dir_path_absolute.startswith("/teamspace/datasets") and len(dir_path_absolute.split("/")) > 3:
+        return _resolve_datasets(dir_path_absolute)
+
+    return Dir(path=dir_path_absolute, url=None)
+
+
+def _match_studio(target_id: Optional[str], target_name: Optional[str], cloudspace: V1CloudSpace) -> bool:
+    if cloudspace.name is not None and target_name is not None and cloudspace.name.lower() == target_name.lower():
+        return True
+
+    if target_id is not None and cloudspace.id == target_id:
+        return True
+
+    if (
+        cloudspace.display_name is not None
+        and target_name is not None
+        and cloudspace.display_name.lower() == target_name.lower()
+    ):
+        return True
+
+    return False
+
+
+def _resolve_studio(dir_path: str, target_name: Optional[str], target_id: Optional[str]) -> Dir:
+    client = LightningClient(retry=False)
```

Comment:
why not retrying?

Reply:
We can retry a little bit but we should retry too much.